src/Servers/Kestrel/Core/src/Internal/Http/ForUpgradeMessageBody.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    /// <summary>
+    /// The upgrade stream uses the raw connection stream instead of going through the RequestBodyPipe. This
+    /// removes the redundant copy from the transport pipe to the body pipe.
+    /// </summary>
+    public class ForUpgrade : Http1MessageBody
+    {
+        public bool _completed;
+        public ForUpgrade(Http1Connection context)
+            : base(context)
+        {
+            RequestUpgrade = true;
+        }
+
+        // This returns IsEmpty so we can avoid draining the body (since it's basically an endless stream)
+        public override bool IsEmpty => true;
+
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            if (_completed)
+            {
+                throw new InvalidOperationException("Reading is not allowed after the reader was completed.");
+            }
+            return _context.Input.ReadAsync(cancellationToken);
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            if (_completed)
+            {
+                throw new InvalidOperationException("Reading is not allowed after the reader was completed.");
+            }
+            return _context.Input.TryRead(out result);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            _context.Input.AdvanceTo(consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            _context.Input.AdvanceTo(consumed, examined);
+        }
+
+        public override void Complete(Exception exception)
+        {
+            // Don't call Connection.Complete.
+            _context.ReportApplicationError(exception);
+            _completed = true;
```

Comment:
I don't believe we should still complete the Connection pipe if the user calls Complete (just like we did with the request side).